test/functional/ui/statusline_spec.lua

```diff
@@ -1,7 +1,69 @@
 local helpers = require('test.functional.helpers')(after_each)
 local Screen = require('test.functional.ui.screen')
-local clear, command, feed = helpers.clear, helpers.command, helpers.feed
-local eq, funcs, meths = helpers.eq, helpers.funcs, helpers.meths
+local clear = helpers.clear
+local command = helpers.command
+local feed = helpers.feed
+local eq = helpers.eq
+local funcs = helpers.funcs
+local meths = helpers.meths
+local exec = helpers.exec
+local eval = helpers.eval
+
+describe('statusline clicks', function()
+  local screen
+
+  before_each(function()
+    clear()
+    screen = Screen.new(40, 8)
+    screen:attach()
+    command('set laststatus=2')
+    exec([=[
+      function! MyClickFunc(minwid, clicks, button, mods)
+        let g:testvar = printf("%d %d %s", a:minwid, a:clicks, a:button)
+      endfunction
+    ]=])
+  end)
+
+  it('works', function()
+    meths.set_option('statusline', 'Not clicky stuff %0@MyClickFunc@Clicky stuff%T')
+    meths.input_mouse('left', 'press', '', 0, 6, 17)
+    eq('0 1 l', eval("g:testvar"))
+    meths.input_mouse('right', 'press', '', 0, 6, 17)
+    eq('0 1 r', eval("g:testvar"))
+  end)
+
+  it('works for winbar', function()
+    meths.set_option('winbar', 'Not clicky stuff %0@MyClickFunc@Clicky stuff%T')
+    meths.input_mouse('left', 'press', '', 0, 0, 17)
+    eq('0 1 l', eval("g:testvar"))
+    meths.input_mouse('right', 'press', '', 0, 6, 17)
+    eq('0 1 r', eval("g:testvar"))
+  end)
+
+  it('works when there\'s multiple windows', function()
+    command('split')
+    meths.set_option('statusline', 'Not clicky stuff %0@MyClickFunc@Clicky stuff%T')
+    meths.set_option('winbar', 'Not clicky stuff %0@MyClickFunc@Clicky stuff%T')
+    meths.input_mouse('left', 'press', '', 0, 0, 17)
+    eq('0 1 l', eval("g:testvar"))
+    meths.input_mouse('right', 'press', '', 0, 4, 17)
+    eq('0 1 r', eval("g:testvar"))
+    meths.input_mouse('middle', 'press', '', 0, 3, 17)
+    eq('0 1 m', eval("g:testvar"))
+    meths.input_mouse('left', 'press', '', 0, 6, 17)
+    eq('0 1 l', eval("g:testvar"))
+  end)
+
+  it('ignores unsupported click items', function()
+    command('tabnew | tabprevious')
+    meths.set_option('statusline', '%2TNot clicky stuff%T')
+    meths.input_mouse('left', 'press', '', 0, 6, 0)
+    eq(1, meths.get_current_tabpage().id)
+    meths.set_option('statusline', '%2XNot clicky stuff%X')
+    meths.input_mouse('left', 'press', '', 0, 6, 0)
+    eq(2, #meths.list_tabpages())
+  end)
+end)
```

Comment:
does a thing like `%0@v:lua.mod.clicky_func@Clicky stuff%T` work? likely it already does, but we could add a test regardless.